OPA Sidecar API changes (#8495)
solo-io/gloo: docs/content/reference/api/github.com/solo-io/gloo/projects/gloo/api/v1/enterprise/options/extauth/v1/extauth.proto.sk.md

| Field | Type | Description |
| ----- | ---- | ----------- | 
| `fastInputConversion` | `bool` | Decreases OPA latency by speeding up conversion of input to the OPA engine. If this is set to true, only http_request and state fields which are a scalar, map, or string array are included in the request input. All other fields are dropped. Dropped fields will not be evaluated by the OPA engine. By default, this is set to false and all fields are evaluated by OPA. |

> | `returnDecisionReason` | `bool` | Set to true to return the reason for an OPA policy decision, based on the logic in your Rego rules. This way, you can use the reason in subsequent filters, such as transformation policies. This `returnDecisionReason` field must be the second parameter of the query when using OpaAuth. When using this field in OpaServerAuth, the entire document will be returned as metadata and the `allowed` field on the document will be used to make the policy decision. The entry will be in the returned DynamicMetadata in the CheckResponse and the structure will be envoy.filters.http.ext_authz: -> name of the auth step, i.e. spec.configs[i].name -> reason If set to false, the response is allowed or denied based on the rego rules without returning the reason. |

Reviewer: There's a period missing
```suggestion
| `returnDecisionReason` | `bool` | Set to true to return the reason for an OPA policy decision, based on the logic in your Rego rules. This way, you can use the reason in subsequent filters, such as transformation policies. This `returnDecisionReason` field must be the second parameter of the query when using OpaAuth. When using this field in OpaServerAuth, the entire document will be returned as metadata and the `allowed` field on the document will be used to make the policy decision. The entry will be in the returned DynamicMetadata in the CheckResponse and the structure will be envoy.filters.http.ext_authz: -> name of the auth step, i.e. spec.configs[i].name -> reason. If set to false, the response is allowed or denied based on the Rego rules without returning the reason. |
```

---
### OpaServerAuth

 
Enforce Open Policy Agent (OPA) policies through an OPA sidecar as part of the external 
auth server in Gloo Platform environments. For Gloo Edge environments, use OpaAuth instead.

```yaml
"package": string
"ruleName": string
"serverAddr": string
"options": .enterprise.gloo.solo.io.OpaAuthOptions

```

| Field | Type | Description |
| ----- | ---- | ----------- | 
| `package` | `string` | The package from your rego policy bundle used to query the OPA data API. |
| `ruleName` | `string` | The rule in your rego policy bundle used to query the OPA data API. Supports querying subfields with a /, see https://www.openpolicyagent.org/docs/latest/rest-api/#data-api. |

> | `serverAddr` | `string` | The address of the OPA server to query, in the format `ADDRESS:PORT`. For OPA servers within the cluster, the address is the pod's service address, such as `default.svc.cluster.local:8181`. For OPA servers outside the cluster, the server must be accessible to the cluster, such as through an ExternalService. If you do not have your own OPA server instance, omit this field. Deploy the ext auth service with the OPA server sidecare enabled and the OPA server sidecar will be used instead. |

Reviewer: typo
```suggestion
| `serverAddr` | `string` | The address of the OPA server to query, in the format `ADDRESS:PORT`. For OPA servers within the cluster, the address is the pod's service address, such as `default.svc.cluster.local:8181`. For OPA servers outside the cluster, the server must be accessible to the cluster, such as through an ExternalService. If you do not have your own OPA server instance, omit this field. Deploy the ext auth service with the OPA server sidecar enabled and the OPA server sidecar will be used instead. |
```

---
### OpaServerAuth

> 

Reviewer: Maybe add a general description to this section to say that it's only for Gloo Platform, based on the discussion in scrum?

> Enforce Open Policy Agent (OPA) policies through an OPA sidecar as part of the external auth server in Gloo Platform environments. For Gloo Edge environments, see [OpaAuth](#opaauth).

Author: OpaServerAuthConfig is specific to gloo and will not be implemented.

```yaml
"package": string
"ruleName": string
"serverAddr": string
"options": .enterprise.gloo.solo.io.OpaAuthOptions

```

| Field | Type | Description |
| ----- | ---- | ----------- | 
| `package` | `string` | The package used to query the OPA data API. |

> | `ruleName` | `string` | The rule used to query the OPA data API. Supports querying subfields with a /, see https://www.openpolicyagent.org/docs/latest/rest-api/#data-api. |

Author: the package and rule is set by the user in their rego bundles. They define and create them. documenting creating these is outside the scope of our API imo so I just linked to the OPA docs - wdyt?

> | `serverAddr` | `string` | The address of the OPA server to query. Should be in the form ADDRESS:PORT, e.g. `default.svc.cluster.local:8181`. If you are not using your own OPA server instance, you should leave this field empty. The OPA server sidecar will be used. |

Author: the sidecar has to be enabled if there is no `serverAddr` set, yes. I've added language to reflect that. It can be outside the cluster, there are no specific requirements other than it is reachable. Whether or not you need an external service would depend on your networking rules.

| `modules` | `map<string, string>` | An optional modules (filename, module content) maps containing modules assist in the resolution of `query`. |
| `query` | `string` | The query that determines the auth decision. The result of this query must be either a boolean or an array with boolean as the first element. A boolean `true` value means that the request will be authorized. Any other value, or error, means that the request will be denied. |
| `options` | [.enterprise.gloo.solo.io.ExtAuthConfig.OpaAuthOptions](../extauth.proto.sk/#opaauthoptions) | Additional Options for Opa Auth configuration. |

---
### OpaAuthOptions

```yaml
"fastInputConversion": bool
"returnDecisionReason": bool

```

| Field | Type | Description |
| ----- | ---- | ----------- | 
| `fastInputConversion` | `bool` | Decreases OPA latency by speeding up conversion of input to the OPA engine. If this is set to true, only http_request and state fields which are a scalar, map, or string array are included in the request input. All other fields are dropped. Dropped fields will not be evaluated by the OPA engine. By default, this is set to false and all fields are evaluated by OPA. |

> | `returnDecisionReason` | `bool` | Return the reason given from the OPA engine after a decision made on this policy. Reason must be the second parameter of the query. The entry will be in the returned DynamicMetadata in the CheckResponse and the structure will be envoy.filters.http.ext_authz: -> name of the auth step, i.e. spec.configs[i].name -> reason. |

Reviewer: ```suggestion
| `returnDecisionReason` | `bool` | Set to true to return the reason for an OPA policy decision, based on the logic in your Rego rules. This way, you can use the reason in subsequent filters, such as transformation policies. This `returnDecisionReason` field must be the second parameter of the query. The entry will be in the returned as DynamicMetadata in the CheckResponse with the structure `envoy.filters.http.ext_authz: -> name of the auth step`, such as `spec.configs[i].name -> reason`. If set to false, the response is allowed or denied based on the rego rules without returning the reason.|
```

---
### OpaServerAuthConfig

```yaml
"package": string
"ruleName": string
"serverAddr": string
"options": .enterprise.gloo.solo.io.ExtAuthConfig.OpaAuthOptions

```

| Field | Type | Description |
| ----- | ---- | ----------- | 

> | `package` | `string` | The package used to query the OPA data API. |

Reviewer: Similar comments as the `OpaServerAuth` section above.

Reviewer: I'm also wondering what the difference between OpaServerAuth and OpaServerAuthConfig is?
